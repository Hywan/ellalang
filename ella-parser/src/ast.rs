use crate::lexer::Token;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    NumberLit(f64),
    BoolLit(bool),
    StringLit(String),
    /// An identifier (e.g. `foo`).
    Identifier(String),
    /// A function call (e.g. `foo(1, bar, baz())`).
    FnCall {
        ident: String,
        args: Vec<Expr>,
    },
    /// A binary expression (e.g. `1+1`).
    Binary {
        lhs: Box<Expr>,
        op: Token,
        rhs: Box<Expr>,
    },
    Unary {
        op: Token,
        arg: Box<Expr>,
    },
    Error,
}

/// Metadata related to variable declarations.
#[derive(Debug, Clone, PartialEq)]
pub struct DeclarationMeta {
    /// *Default:* `None`
    /// Modified in variable resolution pass.
    pub is_captured_in_closure: Option<bool>,
}

impl DeclarationMeta {
    pub fn new() -> Self {
        Self {
            is_captured_in_closure: Some(false),
        }
    }
}

impl Default for DeclarationMeta {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    LetDeclaration {
        meta: DeclarationMeta,
        ident: String,
        initializer: Expr,
    },
    FnDeclaration {
        meta: DeclarationMeta,
        ident: String,
        params: Vec<String>,
        body: Vec<Stmt>,
    },
    Block(Vec<Stmt>),
    ExprStmt(Expr),
    ReturnStmt(Expr),
    Error,
}
