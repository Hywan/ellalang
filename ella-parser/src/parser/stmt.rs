use crate::ast::DeclarationMeta;

use super::*;

impl<'a> Parser<'a> {
    /// Parses a declaration (or statement).
    pub fn parse_declaration(&mut self) -> Stmt {
        match self.current_token {
            Token::Let => self.parse_let_declaration(),
            Token::Fn => self.parse_fn_declaration(),
            _ => self.parse_stmt(),
        }
    }

    /// Parses a statement.
    pub fn parse_stmt(&mut self) -> Stmt {
        match self.current_token {
            Token::Return => self.parse_return_stmt(),
            Token::OpenBrace => self.parse_block_stmt(),
            _ => {
                // expression statement
                let expr = self.parse_expr();
                let stmt = Stmt::ExprStmt(expr);
                self.expect(Token::Semi);
                stmt
            }
        }
    }

    pub fn parse_block_stmt(&mut self) -> Stmt {
        self.expect(Token::OpenBrace);

        let mut body = Vec::new();
        if !self.eat(Token::CloseBrace) {
            loop {
                body.push(self.parse_declaration());

                if self.eat(Token::CloseBrace) {
                    break;
                } else if self.current_token == Token::Eof {
                    self.unexpected();
                    break;
                }
            }
        }

        Stmt::Block(body)
    }

    fn parse_let_declaration(&mut self) -> Stmt {
        self.expect(Token::Let);
        let ident = if let Token::Identifier(ref ident) = self.current_token {
            let ident = ident.clone();
            self.next();
            ident
        } else {
            self.unexpected();
            return Stmt::Error;
        };
        self.expect(Token::Equals);
        let initializer = self.parse_expr();
        self.expect(Token::Semi);
        Stmt::LetDeclaration {
            meta: DeclarationMeta::new(),
            ident,
            initializer,
        }
    }

    fn parse_fn_declaration(&mut self) -> Stmt {
        self.expect(Token::Fn);
        let ident = if let Token::Identifier(ref ident) = self.current_token {
            let ident = ident.clone();
            self.next();
            ident
        } else {
            self.unexpected();
            return Stmt::Error;
        };
        self.expect(Token::OpenParen);
        let mut params = Vec::new();
        if !self.eat(Token::CloseParen) {
            loop {
                params.push(if let Token::Identifier(ref ident) = self.current_token {
                    let ident = ident.clone();
                    self.next();
                    ident
                } else {
                    self.unexpected();
                    return Stmt::Error;
                });

                if self.eat(Token::CloseParen) {
                    break;
                } else if !self.eat(Token::Comma) {
                    self.unexpected();
                    break;
                }
            }
        }

        self.expect(Token::OpenBrace);

        let mut body = Vec::new();
        if !self.eat(Token::CloseBrace) {
            loop {
                body.push(self.parse_declaration());

                if self.eat(Token::CloseBrace) {
                    break;
                }
            }
        }

        Stmt::FnDeclaration {
            meta: DeclarationMeta::new(),
            body,
            ident,
            params,
        }
    }

    fn parse_return_stmt(&mut self) -> Stmt {
        self.expect(Token::Return);
        let expr = self.parse_expr();
        self.expect(Token::Semi);
        Stmt::ReturnStmt(expr)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    fn stmt(source: &str) -> Stmt {
        let source = source.into();
        let ast = Parser::new(&source).parse_declaration();
        assert!(source.has_no_errors());
        ast
    }

    #[test]
    fn test_block_stmt() {
        assert_debug_snapshot!("block-stmt", stmt("{ 1; }"));
        assert_debug_snapshot!("block-stmt-multiple", stmt("{ 1; 2; }"));
        assert_debug_snapshot!("block-stmt-nested", stmt("{ 1; 2; { 3; } }"));
    }

    #[test]
    fn test_let_declaration() {
        assert_debug_snapshot!("let-declaration", stmt("let x = 2;"));
        assert_debug_snapshot!("let-declaration-with-expr", stmt("let x = 1 + 2;"));
    }

    #[test]
    fn test_fn_declaration() {
        assert_debug_snapshot!("fn-declaration", stmt("fn foo() {}"));
        assert_debug_snapshot!("fn-declaration-with-params", stmt("fn foo(a, b, c) {}"));
        assert_debug_snapshot!(
            "fn-declaration-with-params-and-body",
            stmt("fn foo(a, b, c) { a + b + c; }")
        );
    }

    #[test]
    fn test_return_stmt() {
        assert_debug_snapshot!("return-stmt", stmt("return 1;"));
        assert_debug_snapshot!("return-stmt-with-expr", stmt("return 1 + 2;"));
    }
}
