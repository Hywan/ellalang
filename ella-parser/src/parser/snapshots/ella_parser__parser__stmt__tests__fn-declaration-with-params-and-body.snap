---
source: ella-parser/src/parser/stmt.rs
expression: "stmt(\"fn foo(a, b, c) { a + b + c; }\")"
---
FnDeclaration {
    ident: "foo",
    params: [
        "a",
        "b",
        "c",
    ],
    body: [
        ExprStmt(
            Binary {
                lhs: Binary {
                    lhs: Identifier(
                        "a",
                    ),
                    op: Plus,
                    rhs: Identifier(
                        "b",
                    ),
                },
                op: Plus,
                rhs: Identifier(
                    "c",
                ),
            },
        ),
    ],
}
